diskqueue.go

```diff
@@ -395,6 +401,223 @@ func (d *diskQueue) readOne() ([]byte, error) {
+		curFileName := d.fileName(d.readFileNum)
+
+		d.readFile, err = os.OpenFile(curFileName, os.O_RDONLY, 0600)
+		if err != nil {
+			return err
+		}
+
+		d.reader = bufio.NewReader(d.readFile)
+	}
+
+	closeReadFile := func() {
+		d.readFile.Close()
+		d.readFile = nil
+	}
+	defer closeReadFile()
+
+	// read total messages number at the end of the file
+	_, err = d.readFile.Seek(-numFileMsgBytes, 2)
+	if err != nil {
+		return err
+	}
+
+	var totalMessages int64
+	err = binary.Read(d.reader, binary.BigEndian, &totalMessages)
+	if err != nil {
+		return err
+	}
+
+	// update depth with the remaining number of messages
+	d.depth -= totalMessages - d.readMessages
+
+	// we have not finished reading this file
+	if d.readFileNum == d.nextReadFileNum {
+		d.nextReadFileNum++
+		d.nextReadPos = 0
+	}
+
+	d.moveToNextReadFile()
+
+	return nil
+}
+
+// walk through all of the files in the DiskQueue directory
+func (d *diskQueue) walkDiskQueueDir(fn func(fs.DirEntry) error) error {
+	// the directory containing DiskQueue files
+	var mainDir string
+	if d.dataPath == "/" {
+		mainDir = d.dataPath
+	} else {
+		pathArray := strings.Split(d.dataPath, "/")
+		mainDir = pathArray[len(pathArray)-1]
+	}
+
+	walkDiskQueueDir := func(pathStr string, dirEntry fs.DirEntry, err error) error {
+		if dirEntry.Name() == mainDir {
+			// we want to see the contents of the directory DiskQueue writes and reads in
+			return nil
+		}
+
+		if dirEntry.IsDir() {
+			// if the entry is a directory, skip it
+			return fs.SkipDir
+		}
+
+		if err != nil {
+			return err
+		}
+
+		return fn(dirEntry)
+	}
+
+	err := filepath.WalkDir(d.dataPath, walkDiskQueueDir)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *diskQueue) getAllBadFileInfo() ([]fs.FileInfo, error) {
+	var badFileInfos []fs.FileInfo
+
+	getAllBadFileInfo := func(dirEntry fs.DirEntry) error {
+		// only accept "bad" files created by this DiskQueue object
+		regExp, _ := regexp.Compile(`^` + d.name + `.diskqueue.\d\d\d\d\d\d.dat.bad$`)
+
+		if regExp.MatchString(dirEntry.Name()) {
+			badFileInfo, e := dirEntry.Info()
+			if e == nil && badFileInfo != nil {
+				badFileInfos = append(badFileInfos, badFileInfo)
+			}
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(getAllBadFileInfo)
+
+	return badFileInfos, err
+}
+
+// get the accurate total non-"bad" file size
+func (d *diskQueue) updateWriteBytes() error {
+	d.writeBytes = 0
+
+	updateWriteBytes := func(dirEntry fs.DirEntry) error {
+		// only accept files created by this DiskQueue object
+		regExp, _ := regexp.Compile(`^` + d.name + `.diskqueue.\d\d\d\d\d\d.dat$`)
+		if regExp.MatchString(dirEntry.Name()) {
+			fileInfo, e := dirEntry.Info()
+			if e == nil && fileInfo != nil {
+				d.writeBytes += fileInfo.Size()
+			}
+		}
+
+		return nil
+	}
+
+	return d.walkDiskQueueDir(updateWriteBytes)
+}
+
+// free up disk space if needed to write new data to file
+func (d *diskQueue) freeUpDiskSpace(totalBytes int64, reachedFileSizeLimit bool) error {
```

Comment:
`freeUpDiskSpace` is in the path of every `writeOne` and it's scaning the folder for bad files everytime, because `d.WriteBytes` only represents total size of normal data files.  If we use `WriteBytes`(or maybe another better name)  to represent all the files including normal, bad, and metadata, then we can avoid folder scan from each `writeOne` and only do the folder scan to update `WriteBytes` when we're really not sure, e.g. when 1. restart; 2. mark file as bad; (maybe not neccessary 3. after file delete)

Reply:
I love the idea of using one single `totalBytesUsed` or `totalDiskSpaceUsed` variable to account for the total disk size used by DiskQueue. For the folder update, should we still do the folder scan when deleting a read file that was already read too? (similar to #3).

Reply:
> when deleting a read file that was already read

+1 on decreasing `totalDiskSpaceUsed` by the size of the deleted file, since when we delete a file we can get its accurate size, which is more efficient than getting the size of all other files again.